Assets/Scripts/AchievementsService.cs

```diff
@@ -348,34 +353,46 @@ public Texture2D GetAchievementLockedIconTexture(string achievementId)
         /// The icon texture associated with the given achievement ID and
         /// determined by the given URI selector.
         /// </returns>
-        private Texture2D GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
+        private Task<Texture2D> GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
         {
-            Texture2D iconTexture = null;
+            TaskCompletionSource<Texture2D> tcs = new();
             foreach (var achievementDef in _achievements)
             {
                 if (achievementDef.AchievementId != achievementId)
                     continue;
 
-                if (_downloadCache.TryGetValue(uriSelector(achievementDef), out byte[] iconBytes) && null != iconBytes)
+                var uri = uriSelector(achievementDef);
+
+                
+                Texture2D textureFromBytes = new(2, 2);
+
+                if (!_downloadCache.ContainsKey(uri))
                 {
-                    Texture2D textureFromBytes = new(2, 2);
-                    if (textureFromBytes.LoadImage(iconBytes))
-                    {
-                        iconTexture = textureFromBytes;
-                    }
-                    else
+                    GetAndCacheData(uri, data =>
```

Comment:
It feels like we should either exit out of the query now, or use an else to bridge to the next thing.

Currently if the download cache doesn't contain the key, it'll kick off GetAndCacheData. Then it'll see that the data is or isn't in the download cache. Then it'll set the result from bytes, even if that's still the default texture it made earlier.

Reply:
Excellent point! I have updated the implementation and now I believe things are handled more appropriately.